ZiLing/Midterm-Review.py

```diff
@@ -0,0 +1,481 @@
+"""
+Midterm review for cs 114
+
+Instructor: Cameron Morland - cjmorland@uwaterloo.ca
+
+This review will cover materials from module 1 to 6, from all the slides 
+and the textbook materials
+
+"""
+
+"""
+Let's start with the very basic
+
+Types         name                        examples 
+
+int          Interger                  3,4,-1,2325289524294
+float   Float-Point Numbers            4.5,0.0,-4305.2458532
+str          String                   'hello', "world", " ",''
+bool         Boolean                   True,False, 1 == 1, 1 != 1
+list          List                     [], [12,3.0432,'hello'],['a','b','v']
+tuple         tuple                    (True,True,False),(),(23,43.2,"Tuple")
+dict        Dictionary                 {2:4},{454:'ok','ok':'No'}
+range         Range                    range(4),range(1,4302,45)
+callable     Callable                  print(),count_To_Four(), f()
+None         NoneType                          None
+Further explination for each will be explained further later, now just note that 
+these are all the types we are working with  
+"""
```

Comment:
Note about int, float, and str in Python.

In C, you can directly control the size of an int or float (technically this is abstracted away and depends on how the CPU does it, but every CPU does it the same way now), and chars are of fixed size (technically no but something similar as what applies to floats).

An integer can be 8, 16, 32, or 64 bits, the latter being referred to as a "long." There are 8 bits to a byte with every modern CPU. A float is either 32 bits or 64 bits, the latter being referred to as a double.  A char is always one byte, or eight bits.

To understand why the size matters, we need to look at what the size restricts. The number of values that an int can take on is two raised to the power of its bit size. So an 8-bit integer can take on 256 values. Here is where integers can again vary, which is explicit in C: they can be unsigned or signed. For a signed integer, it can be negative or non-negative. So an 8-bit signed integer takes on values of -127 to 128. On the other hand, an unsigned 8-bit integer can take on values of 0 to 255. 

For floats, the bit size determines the number of decimal points you have for precision. A 32-bit bloat has 7 decimal points of precision. A double (64-bit float) has double that precision at 14 decimal points, hence the name.

Python, despite being written on top of C, does not use this convention and its ints and floats have variable bit sizes, with floats extending up to 53 bits with 15 decimal points of precision. This is both good and bad. On the one hand, you can avoid worrying about bit sizes of your types. It's either small enough or it isn't and Python will automatically use a compatable bit-size However, this can make your program run slower, because Python isn't smart enough to use the minimally sized object to store the value of your object. Thus, you can end up with a 64-bit integer for `integer = 5` even though you can use an 8-bit integer. The larger an object it is, the more memory it takes up and the slower reading its value is. This is also problematic if you are writing to binary files (pretty much every modern machine uses binary for input-output [IO or I/O]) since you need to know how big each object is.

Numpy allows you to use c-types. See the following documentation

[calling c-types](https://numpy.org/devdocs/user/basics.types.html)
[scalars](https://numpy.org/devdocs/reference/arrays.scalars.html#numpy.int_)